LibraBFT/Impl/Consensus/SafetyRules/SafetyRules.agda

```diff
@@ -145,6 +151,65 @@ verifyQcM qc = do
 
 ------------------------------------------------------------------------------
 
+consensusState : SafetyRules → ConsensusState
+consensusState self =
+  ConsensusState∙new (self ^∙ srPersistentStorage ∙ pssSafetyData)
+                     (self ^∙ srPersistentStorage ∙ pssWaypoint)
+                     true -- TODO
+
+------------------------------------------------------------------------------
+
+-- ** NOTE: PAY PARTICULAR ATTENTION TO THIS FUNCTION **
+initialize : SafetyRules → EpochChangeProof → Either ErrLog SafetyRules
+initialize self proof = do
+  let waypoint   = self ^∙ srPersistentStorage ∙ pssWaypoint
+  lastLi         ← withErrCtx' (here' ("EpochChangeProof.verify" ∷ []))
+                               (        EpochChangeProof.verify proof waypoint)
+  let ledgerInfo = lastLi ^∙ liwsLedgerInfo
+  epochState     ← maybeS (ledgerInfo ^∙ liNextEpochState)
+                          (Left fakeErr) -- ["last ledger info has no epoch state"]
+                          pure
+  let currentEpoch = self ^∙ srPersistentStorage ∙ pssSafetyData ∙ sdEpoch
+  if-dec currentEpoch <? epochState ^∙ esEpoch
+    then (do
+      waypoint'  ← withErrCtx' (here' ("Waypoint.newEpochBoundary" ∷ []))
+                               (        Waypoint.newEpochBoundary ledgerInfo)
+      continue1 (self & srPersistentStorage ∙ pssWaypoint    ∙~ waypoint'
+                      & srPersistentStorage ∙ pssSafetyData  ∙~
+                          SafetyData∙new (epochState ^∙ esEpoch) {-Round-} 0 {-Round-} 0 nothing)
+                epochState)
+    else continue1 self epochState
+ where
+  here'     : List String.String → List String.String
+  continue2 : SafetyRules → EpochState → Either ErrLog SafetyRules
+  continue1 : SafetyRules → EpochState → Either ErrLog SafetyRules
+
+  continue1 self1 epochState =
+    continue2 (self1 & srEpochState ?~ epochState) epochState
+
+  continue2 self2 epochState = do
+    let author = self2 ^∙ srPersistentStorage ∙ pssAuthor
+    maybeS (ValidatorVerifier.getPublicKey (epochState ^∙ esVerifier) author)
+      (Left fakeErr) -- ["ValidatorNotInSet", lsA author] $
+      λ expectedKey → do
+        let currKey = eitherS (signer self2)
+                      (const nothing)
+                      (just ∘ ValidatorSigner.publicKey_ONLY_USE_AT_INIT)
```

Comment:
Style police: should be spelled `publicKey_USE_ONLY_AT_INIT`.

Reply:
DONE